File: script/make/protocol/lua/NoticeProtocol.lua
NoticeProtocol = {}

function NoticeProtocol.encode(offset, protocol, data)
    if protocol == 50001 then
        local offset = offset
        local table = {}
        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function NoticeProtocol.decode(offset, protocol, data)
    if protocol == 50001 then
        local offset = offset
        -- 公告列表
        local noticeList = {}
        local noticeListLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for noticeListIndex = 1, noticeListLength do
            -- NoticeRole
            -- 公告ID
            local noticeId = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 收到时间
            local receiveTime = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 读取时间
            local readTime = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 标题
            local title = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(title)
            -- 内容
            local content = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(content)
            -- object
            local noticeRole = {noticeId = noticeId, receiveTime = receiveTime, readTime = readTime, title = title, content = content}
            noticeList[noticeListIndex] = noticeRole
        end
        return {noticeList = noticeList}
    elseif protocol == 50002 then
        local offset = offset
        -- 范围
        local scope = string.unpack(">I1", data, offset)
        offset = offset + 1
        -- 类型
        local type = string.unpack(">I1", data, offset)
        offset = offset + 1
        -- 标题
        local title = string.unpack(">s2", data, offset)
        offset = offset + 2 + string.len(title)
        -- 消息
        local msg = string.unpack(">s2", data, offset)
        offset = offset + 2 + string.len(msg)
        return {scope = scope, type = type, title = title, msg = msg}
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

File: script/make/protocol/cs/NoticeProtocol.cs
public static class NoticeProtocol
{
    public static void Encode(System.Text.Encoding encoding, System.IO.BinaryWriter writer, System.UInt16 protocol, System.Object data) 
    {
        switch (protocol) 
        {
            case 50001:
            {
                return;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }

    public static System.Object Decode(System.Text.Encoding encoding, System.IO.BinaryReader reader, System.UInt16 protocol) 
    {
        switch (protocol) 
        {
            case 50001:
            {
                // 公告列表
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = new System.Collections.Generic.List<System.Object>(dataLength);
                while (dataLength-- > 0)
                {
                    // 
                    // 公告ID
                    var noticeId = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 收到时间
                    var receiveTime = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 读取时间
                    var readTime = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 标题
                    var titleLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var title = encoding.GetString(reader.ReadBytes(titleLength));
                    // 内容
                    var contentLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var content = encoding.GetString(reader.ReadBytes(contentLength));
                    // object
                    var noticeRole = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"noticeId", noticeId}, {"receiveTime", receiveTime}, {"readTime", readTime}, {"title", title}, {"content", content}};
                    // add
                    data.Add(noticeRole);
                }
                return data;
            }
            case 50002:
            {
                // 
                // 范围
                var scope = reader.ReadByte();
                // 类型
                var type = reader.ReadByte();
                // 标题
                var titleLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var title = encoding.GetString(reader.ReadBytes(titleLength));
                // 消息
                var msgLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var msg = encoding.GetString(reader.ReadBytes(msgLength));
                // object
                var data = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"scope", scope}, {"type", type}, {"title", title}, {"msg", msg}};
                return data;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }
}